### ISSUE TITLE
Memory leak in ICMP ping (IDFGH-6489)

### ISSUE THREAD

**Issue by MaazSk** [espressif/esp-idf#8143]
## Environment

- Development Kit:      [ESP32-DevKit]
- Kit version (for WroverKit/PicoKit/DevKitC): [v1]
- Module or chip used:  [ESP32-WROOM-32|]
- IDF version (run ``git describe --tags`` to find it): v4.3
   
- Compiler version : xtensa-esp32-elf-gcc (crosstool-NG esp-2021r1) 8.4.0
  
- Operating System:     [Linux]

- Using an IDE?: [VS Code]
- Power Supply:         [USB]

## Problem Description

I was trying the ICMP echo example and there is a possible memory leak in the code.

The code which I am trying is as below:

/* ICMP echo example

   This example code is in the Public Domain (or CC0 licensed, at your option.)

   Unless required by applicable law or agreed to in writing, this
   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied.
*/

```
#include <stdio.h>
#include <string.h>

#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "freertos/FreeRTOSConfig.h"
#include "sdkconfig.h"
#include "lwip/inet.h"
#include "lwip/netdb.h"
#include "lwip/sockets.h"
#include "esp_console.h"
#include "esp_event.h"
#include "nvs_flash.h"
#include "argtable3/argtable3.h"
#include "protocol_examples_common.h"
#include "ping/ping_sock.h"
#include "esp_heap_trace.h"

#define NUM_RECORDS 200
static heap_trace_record_t trace_record[200];

#define portTICK_PERIOD_MS 1

static void ping_app(void *params);

static void cmd_ping_on_ping_success(esp_ping_handle_t hdl, void *args)
{
    uint8_t ttl;
    uint16_t seqno;
    uint32_t elapsed_time, recv_len;
    ip_addr_t target_addr;
    esp_ping_get_profile(hdl, ESP_PING_PROF_SEQNO, &seqno, sizeof(seqno));
    esp_ping_get_profile(hdl, ESP_PING_PROF_TTL, &ttl, sizeof(ttl));
    esp_ping_get_profile(hdl, ESP_PING_PROF_IPADDR, &target_addr, sizeof(target_addr));
    esp_ping_get_profile(hdl, ESP_PING_PROF_SIZE, &recv_len, sizeof(recv_len));
    esp_ping_get_profile(hdl, ESP_PING_PROF_TIMEGAP, &elapsed_time, sizeof(elapsed_time));
    printf("%d bytes from %s icmp_seq=%d ttl=%d time=%d ms\n",
           recv_len, ipaddr_ntoa((ip_addr_t *)&target_addr), seqno, ttl, elapsed_time);
}

static void cmd_ping_on_ping_timeout(esp_ping_handle_t hdl, void *args)
{
    uint16_t seqno;
    ip_addr_t target_addr;
    esp_ping_get_profile(hdl, ESP_PING_PROF_SEQNO, &seqno, sizeof(seqno));
    esp_ping_get_profile(hdl, ESP_PING_PROF_IPADDR, &target_addr, sizeof(target_addr));
    printf("From %s icmp_seq=%d timeout\n", ipaddr_ntoa((ip_addr_t *)&target_addr), seqno);
}

static void cmd_ping_on_ping_end(esp_ping_handle_t hdl, void *args)
{
    ip_addr_t target_addr;
    uint32_t transmitted;
    uint32_t received;
    uint32_t total_time_ms;
    esp_ping_get_profile(hdl, ESP_PING_PROF_REQUEST, &transmitted, sizeof(transmitted));
    esp_ping_get_profile(hdl, ESP_PING_PROF_REPLY, &received, sizeof(received));
    esp_ping_get_profile(hdl, ESP_PING_PROF_IPADDR, &target_addr, sizeof(target_addr));
    esp_ping_get_profile(hdl, ESP_PING_PROF_DURATION, &total_time_ms, sizeof(total_time_ms));
    uint32_t loss = (uint32_t)((1 - ((float)received) / transmitted) * 100);
    if (IP_IS_V4(&target_addr))
    {
        printf("\n--- %s ping statistics ---\n\n", inet_ntoa(*ip_2_ip4(&target_addr)));
    }
    else
    {
        printf("\n--- %s ping statistics ---\n\n", inet6_ntoa(*ip_2_ip6(&target_addr)));
    }
    printf("%d packets transmitted, %d received, %d%% packet loss, time %dms\n\n",
           transmitted, received, loss, total_time_ms);
}

// static struct
// {
//     struct arg_dbl *timeout;
//     struct arg_dbl *interval;
//     struct arg_int *data_size;
//     struct arg_int *count;
//     struct arg_int *tos;
//     struct arg_str *host;
//     struct arg_end *end;
// } ping_args;

void app_main(void)
{
    ESP_ERROR_CHECK(heap_trace_init_standalone(trace_record, NUM_RECORDS));
    ESP_ERROR_CHECK(nvs_flash_init());
    ESP_ERROR_CHECK(esp_netif_init());
    ESP_ERROR_CHECK(esp_event_loop_create_default());
    /* wait for active network connection */
    ESP_ERROR_CHECK(example_connect());

    xTaskCreate(&ping_app, "Ping", 2048, NULL, 5, NULL);
}

static void ping_app(void *params)
{
    ESP_ERROR_CHECK(heap_trace_start(HEAP_TRACE_LEAKS));
    esp_ping_config_t config = ESP_PING_DEFAULT_CONFIG();
    ip_addr_t target_addr;

    struct addrinfo hint;
    struct addrinfo *res = NULL;
    memset(&hint, 0, sizeof(hint));
    /* convert ip4 string or hostname to ip4 or ip6 address */
    if (getaddrinfo("google.com", NULL, &hint, &res) != 0)
    {
        printf("ping: unknown host %s\n", "google.com");
        return 1;
    }
    if (res->ai_family == AF_INET)
    {
        struct in_addr addr4 = ((struct sockaddr_in *)(res->ai_addr))->sin_addr;
        inet_addr_to_ip4addr(ip_2_ip4(&target_addr), &addr4);
    }
    else
    {
        struct in6_addr addr6 = ((struct sockaddr_in6 *)(res->ai_addr))->sin6_addr;
        inet6_addr_to_ip6addr(ip_2_ip6(&target_addr), &addr6);
    }
    freeaddrinfo(res);

    config.target_addr = target_addr;

    esp_ping_callbacks_t cbs = {
        .on_ping_success = cmd_ping_on_ping_success,
        .on_ping_timeout = cmd_ping_on_ping_timeout,
        .on_ping_end = cmd_ping_on_ping_end,
        .cb_args = NULL};

    esp_ping_handle_t ping;
    esp_ping_new_session(&config, &cbs, &ping);

    // xTaskCreate(ping_apps, "Ping", 2048, NULL, 5, NULL);
    // TickType_t xLastWakeTime;
    // const TickType_t xFrequency = 10;

    // // Initialise the xLastWakeTime variable with the current time.
    // xLastWakeTime = xTaskGetTickCount();
    while (1)
    {

        esp_ping_start(ping);

        heap_trace_dump();
        esp_ping_stop(ping);
        esp_ping_delete_session(ping);
        ESP_ERROR_CHECK(heap_trace_stop());
        vTaskDelay(60000 / portTICK_PERIOD_MS);
    }
}
```

The heap trace logs are as below

> 14 allocations trace (200 entry buffer)
> 
> --- 216.58.203.46 ping statistics ---
> 
> 4 bytes (@ 0x3ffc8208) allocated CPU 1 ccount 0x313a9594 caller 0x400e3ab4:0x400f3718
> 0x400e3ab4: mem_malloc at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/mem.c:237
> 
> 0x400f3718: sys_thread_sem_init at /home/maaz/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:523
> 
> 84 bytes (@ 0x3ffbae0c) allocated CPU 1 ccount 0x313aa00c caller 0x40089312:0x400f373f
> 0x40089312: xQueueGenericCreate at /home/maaz/esp/esp-idf/components/freertos/queue.c:413
> 
> 0x400f373f: sys_thread_sem_init at /home/maaz/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:530
> 
> 4 bytes (@ 0x3ffbae64) allocated CPU 1 ccount 0x313ab3bc caller 0x400d302f:0x400f376d
> 0x400d302f: pthread_setspecific at /home/maaz/esp/esp-idf/components/pthread/pthread_local_storage.c:222
> 
> 0x400f376d: sys_thread_sem_init at /home/maaz/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:539
> 
> 12 bytes (@ 0x3ffbae74) allocated CPU 1 ccount 0x313ac870 caller 0x400d307f:0x400f376d
> 0x400d307f: pthread_setspecific at /home/maaz/esp/esp-idf/components/pthread/pthread_local_storage.c:248
> 
> 0x400f376d: sys_thread_sem_init at /home/maaz/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:539
> 
> 120 bytes (@ 0x3ffc888c) allocated CPU 0 ccount 0x34de9aac caller 0x40084b04:0x40084b18
> 0x40084b04: wifi_calloc at /home/maaz/esp/esp-idf/components/esp_wifi/esp32/esp_adapter.c:107
> 
> 0x40084b18: wifi_zalloc_wrapper at /home/maaz/esp/esp-idf/components/esp_wifi/esp32/esp_adapter.c:113
> 
> 120 bytes (@ 0x3ffc86c8) allocated CPU 0 ccount 0x35018420 caller 0x400e3b24:0x400f4525
> 0x400e3b24: mem_calloc at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/mem.c:1013
> 
> 0x400f4525: esp_ping_new_session at /home/maaz/esp/esp-idf/components/lwip/apps/ping/ping_sock.c:222 (discriminator 2)
> 
> 2048 bytes (@ 0x3ffbb240) allocated CPU 0 ccount 0x350193bc caller 0x4008a1cb:0x400f456f
> 0x4008a1cb: xTaskCreatePinnedToCore at /home/maaz/esp/esp-idf/components/freertos/tasks.c:830
> 
> 0x400f456f: esp_ping_new_session at /home/maaz/esp/esp-idf/components/lwip/apps/ping/ping_sock.c:231 (discriminator 2)
> 
> 344 bytes (@ 0x3ffc8908) allocated CPU 0 ccount 0x35019b08 caller 0x4008a1d9:0x400f456f
> 0x4008a1d9: xTaskCreatePinnedToCore at /home/maaz/esp/esp-idf/components/freertos/tasks.c:835
> 
> 0x400f456f: esp_ping_new_session at /home/maaz/esp/esp-idf/components/lwip/apps/ping/ping_sock.c:231 (discriminator 2)
> 
> 72 bytes (@ 0x3ffc8744) allocated CPU 0 ccount 0x3501bc68 caller 0x400e3b24:0x400f45cc
> 0x400e3b24: mem_calloc at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/mem.c:1013
> 
> 0x400f45cc: esp_ping_new_session at /home/maaz/esp/esp-idf/components/lwip/apps/ping/ping_sock.c:245
> 
> 56 bytes (@ 0x3ffc8790) allocated CPU 0 ccount 0x3501e32c caller 0x400e3ab4:0x400e3b3e
> 0x400e3ab4: mem_malloc at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/mem.c:237
> 
> 0x400e3b3e: do_memp_malloc_pool at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/memp.c:254
> 
> 8 bytes (@ 0x3ffbea80) allocated CPU 0 ccount 0x3501f484 caller 0x400e3ab4:0x400f352d
> 0x400e3ab4: mem_malloc at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/mem.c:237
> 
> 0x400f352d: sys_mbox_new at /home/maaz/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:217
> 
> 0 packets transmitted, 0 received, -1% packet loss, time 0ms
> 
> 108 bytes (@ 0x3ffc87cc) allocated CPU 0 ccount 0x3501fe7c caller 0x40089312:0x400f353d
> 0x40089312: xQueueGenericCreate at /home/maaz/esp/esp-idf/components/freertos/queue.c:413
> 
> 0x400f353d: sys_mbox_new at /home/maaz/esp/esp-idf/components/lwip/port/esp32/freertos/sys_arch.c:223
> 
> 72 bytes (@ 0x3ffc883c) allocated CPU 0 ccount 0x35025c8c caller 0x400e3ab4:0x400e3b3e
> 0x400e3ab4: mem_malloc at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/mem.c:237
> 
> 0x400e3b3e: do_memp_malloc_pool at /home/maaz/esp/esp-idf/components/lwip/lwip/src/core/memp.c:254
> 
> 84 bytes (@ 0x3ffc8a64) allocated CPU 0 ccount 0x3502d148 caller 0x40089312:0x400895af
> 0x40089312: xQueueGenericCreate at /home/maaz/esp/esp-idf/components/freertos/queue.c:413
> 
> 0x400895af: xQueueCreateMutex at /home/maaz/esp/esp-idf/components/freertos/queue.c:525
> 
> 3136 bytes 'leaked' in trace (14 allocations)
> total allocations 32 total frees 18

I can see from these logs that 3408 bytes are been leaked and my code calls it every minute.

**Comment by ESP-YJM** [comment]
@MaazSk I think your approach is not very correct, because `esp_ping` runs as a task. Although you have executed `esp_ping_stop(ping)` and `esp_ping_delete_session(ping)`, the task of `esp_ping` will not exit immediately, and the resources will not be recovered immediately, so there will be the memory leaks you see.

**Comment by MaazSk** [comment]
@ESP-YJM Thank you for your reply, I discovered the same yesterday only and I have already made the changes you have suggested. But still, I don't know why when my device is running for a long time I am getting this error.

```
***ERROR*** A stack overflow in task Ping has been detected.

Backtrace:0x40088f4b:0x3ffcae80 0x40089749:0x3ffcaea0 0x4008d0e1:0x3ffcaec0 0x4008b622:0x3ffcaf40 0x40089840:0x3ffcaf60 0x400897f2:0x00000000 |<-CORRUPTED

```

Can you please help me with this? The stack overflow happened after 4-5 hours from the time the device first booted

**Comment by ESP-YJM** [comment]
@MaazSk You can simply en-large the ping task size by `config.task_stack_size = 4096`.

**Comment by MaazSk** [comment]
Ok, I will try and let you know the output.

Thank you! @ESP-YJM

**Comment by MaazSk** [comment]
@ESP-YJM I checked just now changing the stack size to 4096 but still I am getting the stack overflow error as below

> ***ERROR*** A stack overflow in task ping has been detected.
> 
> Backtrace:0x40088f4b:0x3ffcee30 0x40089749:0x3ffcee50 0x4008d0e1:0x3ffcee70 0x4008b622:0x3ffceef0 0x40089840:0x3ffcef10 0x400897f2:0x58575655 |<-CORRUPTED
> 0x40088f4b: panic_abort at /home/maaz/esp/esp-idf/components/esp_system/panic.c:368
> 
> 0x40089749: esp_system_abort at /home/maaz/esp/esp-idf/components/esp_system/system_api.c:112
> 
> 0x4008d0e1: vApplicationStackOverflowHook at /home/maaz/esp/esp-idf/components/freertos/port/xtensa/port.c:490
> 
> 0x4008b622: vTaskSwitchContext at /home/maaz/esp/esp-idf/components/freertos/tasks.c:3289
> 
> 0x40089840: _frxt_dispatch at /home/maaz/esp/esp-idf/components/freertos/port/xtensa/portasm.S:432
> 
> 0x400897f2: _frxt_int_exit at /home/maaz/esp/esp-idf/components/freertos/port/xtensa/portasm.S:231

**Comment by ESP-YJM** [comment]
@MaazSk Could you please offer the code that i can reproduce locally.

**Comment by MaazSk** [comment]
@ESP-YJM, I think it is a bug in my code let me confirm it first and then I will share the snippet because the standalone code of ping is working fine. Just give me a day I will keep you updated.

Thank you!

**Comment by MaazSk** [comment]
There was a bug in my code sorry.
